File: kernel/src/arch/x86_64/paging.zig
//! Paging structures for x86_64

const std = @import("std");
const Error = std.mem.Allocator.Error;
const root = @import("root");
const logger = std.log.scoped(.x86_64_internals_paging);

pub const PageTable = [512]PageTableEntry;

// TODO Make this a bit field struct
pub const PageTableEntry = packed struct {
    __data: u64,

    pub fn fromU64(entry: u64) PageTableEntry {
        return PageTableEntry{ .__data = entry };
    }

    pub fn isPresent(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000001 != 0;
    }

    pub fn isWritable(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000002 != 0;
    }

    pub fn isUserAccessable(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000004 != 0;
    }

    pub fn writeThroughCachingEnabled(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000008 != 0;
    }

    pub fn cacheDisabled(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000010 != 0;
    }

    pub fn isAccessed(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000020 != 0;
    }

    pub fn isDirty(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000040 != 0;
    }

    pub fn isHugePage(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000080 != 0;
    }

    pub fn isGlobal(self: PageTableEntry) bool {
        return self.__data & 0x0000000000000100 != 0;
    }

    pub fn isNoExecute(self: PageTableEntry) bool {
        return self.__data & 0x8000000000000000 != 0;
    }

    pub fn getKernelData1(self: PageTableEntry) u3 {
        return @truncate(u3, (self.__data & 0xE00) >> 9);
    }

    pub fn setKernelData1(self: *PageTableEntry, data: u3) void {
        self.__data = (self.__data & ~@as(u64, 0xE00)) | (@as(u64, data) << 9);
    }

    pub fn getKernelData2(self: PageTableEntry) u7 {
        return @truncate(u7, (self.__data & 0x7F0000000000000) >> 52);
    }

    pub fn setKernelData2(self: *PageTableEntry, data: u7) void {
        self.__data = (self.__data & ~@as(u64, 0x7F0000000000000)) | (@as(u64, data) << 52);
    }

    pub const framebuffer_flags: u64 = 0x8000000000000083;

    pub const InputFlags = struct {
        present: bool = false,
        writable: bool = false,
        user_accessible: bool = false,
        write_through_caching_enabled: bool = false,
        cache_disabled: bool = false,
        accessed: bool = false,
        dirty: bool = false,
        huge_page: bool = false,
        global: bool = false,
        physical_address: usize = 0,
        no_execute: bool = false,
    };

    pub fn generateU64(flags: InputFlags) u64 {
        return @as(u64, 0) |
            @as(u64, if (flags.present) 0x1 else 0) |
            @as(u64, if (flags.writable) 0x2 else 0) |
            @as(u64, if (flags.user_accessible) 0x4 else 0) |
            @as(u64, if (flags.write_through_caching_enabled) 0x8 else 0) |
            @as(u64, if (flags.cache_disabled) 0x10 else 0) |
            @as(u64, if (flags.accessed) 0x20 else 0) |
            @as(u64, if (flags.dirty) 0x40 else 0) |
            @as(u64, if (flags.huge_page) 0x80 else 0) |
            @as(u64, if (flags.global) 0x100 else 0) |
            @as(u64, if (flags.no_execute) 0x8000_0000_0000_0000 else 0) |
            @as(u64, flags.physical_address & 0x000FFFFFFFFFF000);
    }

    pub fn getAddress(self: PageTableEntry) u64 {
        const addr = self.__data & 0x000FFFFFFFFFF000;
        if (addr & 0x0008000000000000 != 0) {
            return addr | 0xFFF0000000000000;
        } else {
            return addr;
        }
    }
};

/// Translates a virtual address to a physical address
pub fn translateAddress(address: u64) ?u64 {
    // Query memory map
    const mem_map = PageTableEntry.fromU64(asm ("movq %%cr3, %[ret]"
        : [ret] "=r" (-> usize)
    ));
    const level_masks = [_]u64{
        0xFF80_0000_0000,
        0x007F_C000_0000,
        0x0000_3FE0_0000,
        0x0000_001F_F000,
    };
    var current_address = mem_map.getAddress();
    for (level_masks) |level_mask, i| {
        const current_table = @intToPtr(*align(4096) PageTable, current_address);
        const index = @truncate(
            u9,
            (level_mask & address) >> @truncate(u6, (3 - i) * 9 + 12),
        );
        const entry = &current_table[index];
        if (!entry.isPresent()) {
            return null;
        }
        if (entry.isHugePage()) {
            @panic("huge pages not supported");
        }
        current_address = entry.getAddress();
    }
    return current_address | (address & 0xFFF);
}

/// Aligns the address down to the nearest page boundary
pub inline fn alignToPage(address: usize) usize {
    return address & ~@as(usize, 0xFFF);
}

File: kernel_rust/src/arch/x86_64/paging.rs
//! Implementation of x86_64 page tables.

pub const PAGE_SIZE: usize = 4096;

/// Aligns `address` down to the nearest page boundary.
#[inline]
pub fn align_to_page(address: usize) -> usize {
    return address & !0xFFF;
}

pub type PageTable = [PageTableEntry; 512];

bitfield::bitfield! {
    #[derive(Clone, Copy, PartialEq, Eq)]
    #[repr(transparent)]
    pub struct PageTableEntry(u64);
    impl Debug;
    pub present, _: 0;
    pub writable, _: 1;
    pub user_accessable, _: 2;
    pub write_through_caching_enabled, _: 3;
    pub cache_disabled, _: 4;
    pub accessed, _: 5;
    pub dirty, _: 6;
    pub huge_page, _: 7;
    pub global, _: 8;
    pub no_execute, _: 63;
    address_unextended, _: 51, 12;
    kernel_data_1, _: 11, 9;
    kernel_data_2, _: 58, 52;
}

impl PageTableEntry {
    pub const ZERO: Self = Self(0);
    pub const READ: Self = Self::from_data(PageTableData {
        present: true,
        writable: false,
        user_accessable: false,
        write_through_caching_enabled: false,
        cache_disabled: false,
        accessed: false,
        dirty: false,
        huge_page: false,
        global: false,
        physical_address: 0,
        no_execute: true,
    });
    pub const READ_WRITE: Self = Self::from_data(PageTableData {
        present: true,
        writable: true,
        user_accessable: false,
        write_through_caching_enabled: false,
        cache_disabled: false,
        accessed: false,
        dirty: false,
        huge_page: false,
        global: false,
        physical_address: 0,
        no_execute: true,
    });
    pub const READ_EXECUTE: Self = Self::from_data(PageTableData {
        present: true,
        writable: false,
        user_accessable: false,
        write_through_caching_enabled: false,
        cache_disabled: false,
        accessed: false,
        dirty: false,
        huge_page: false,
        global: false,
        physical_address: 0,
        no_execute: false,
    });
    pub const READ_WRITE_EXECUTE: Self = Self::from_data(PageTableData {
        present: true,
        writable: true,
        user_accessable: false,
        write_through_caching_enabled: false,
        cache_disabled: false,
        accessed: false,
        dirty: false,
        huge_page: false,
        global: false,
        physical_address: 0,
        no_execute: false,
    });

    pub const fn from_data(data: PageTableData) -> Self {
        Self(
            data.present as u64
                | (data.writable as u64) << 1
                | (data.user_accessable as u64) << 2
                | (data.write_through_caching_enabled as u64) << 3
                | (data.cache_disabled as u64) << 4
                | (data.accessed as u64) << 5
                | (data.dirty as u64) << 6
                | (data.huge_page as u64) << 7
                | (data.global as u64) << 8
                | (data.no_execute as u64) << 63
                | (data.physical_address as u64 & 0x000FFFFFFFFFF000),
        )
    }

    pub fn address(&self) -> usize {
        let addr_unextended = self.address_unextended() << 12;
        if addr_unextended & 0x0008000000000000 != 0 {
            addr_unextended as usize | 0xFFF0000000000000
        } else {
            addr_unextended as usize
        }
    }

    #[must_use]
    pub const fn replace_flags_with(&self, flags: PageTableEntry) -> Self {
        let raw_address = self.0 & 0x000FFFFFFFFFF000;
        let raw_flags = flags.0 & 0x80000000000001FF;
        Self(raw_address | raw_flags)
    }

    #[must_use]
    pub const fn replace_addr_with(&self, addr: usize) -> Self {
        let stripped_address = addr as u64 & 0x000FFFFFFFFFF000;
        let raw_flags = self.0 & 0x80000000000001FF;
        Self(stripped_address | raw_flags)
    }
}

impl From<PageTableData> for PageTableEntry {
    fn from(data: PageTableData) -> Self {
        Self::from_data(data)
    }
}

impl Default for PageTableEntry {
    fn default() -> Self {
        Self::from(PageTableData::default())
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct PageTableData {
    pub present: bool,
    pub writable: bool,
    pub user_accessable: bool,
    pub write_through_caching_enabled: bool,
    pub cache_disabled: bool,
    pub accessed: bool,
    pub dirty: bool,
    pub huge_page: bool,
    pub global: bool,
    pub physical_address: usize,
    pub no_execute: bool,
}

impl Default for PageTableData {
    fn default() -> Self {
        Self {
            present: true,
            writable: false,
            user_accessable: false,
            write_through_caching_enabled: false,
            cache_disabled: false,
            accessed: false,
            dirty: false,
            huge_page: false,
            global: false,
            physical_address: 0,
            no_execute: false,
        }
    }
}
